File: clj-examples/threadring/threadring.clj
;; The Computer Language Benchmarks Game
;; http://shootout.alioth.debian.org/
;;
;; contributed by Kenneth Jonsson

(ns threadring.threadring (:gen-class))

(defn pass-token [slot value]
  (if (zero? value)
    (do (println (:id slot))
        ;; (System/exit 0)
        )
    (do
      (send (:next slot) pass-token (dec value))
      slot)))

(defn pass-token-orig [ slot value ]
  (when (zero? value)
    (println (:id slot))
    (System/exit 0))
  (send (:next slot) pass-token (dec value))
  slot)

(defn create-ring-and-start [ _ ring-sz initial-value ]
  ;; "send" will be defered until the state of the current agent has
  ;; been set to the state returned by this function
  (send *agent* pass-token initial-value)
  ;; create a ring of "ring-sz" agents linked via ":next"
  { :id 1
   :next (reduce (fn [ next-slot id ]
                   (agent { :next next-slot :id id }))
                 *agent*
                 (range ring-sz 1 -1)) })

(defn -main [ & args ]
  (send (agent nil)
        create-ring-and-start
        503
        (if (empty? args) 1000 (Integer/parseInt (first args)))))

File: go-lightly-examples/go-examples/src/threadring/threadring.go
/* The Computer Language Benchmarks Game
   http://shootout.alioth.debian.org/

   contributed by KP
*/

package main

import (
   "fmt"
   "os"
   "runtime"
   "strconv"
   "sync"
)

type Token int

type T struct {
   next  *T
   label int
   value int
   mux   sync.Mutex
}

func (w *T) put(v int) {
   w.value = v
   if v == 0 {
      res <- w.label
   } else {
      w.mux.Unlock()
   }
}

func (w *T) run() {
   for {
      w.mux.Lock()
      w.next.put(w.value - 1)
      runtime.Gosched()
   }
}

func (w *T) Start(label int, next *T) {
   w.label = label
   w.next = next
   w.mux.Lock()
   go w.run()
}

const NThreads = 503

var res = make(chan int)

func main() {
   n := 1000
   if len(os.Args) > 1 {
      n, _ = strconv.Atoi(os.Args[1])
   }

   var channels [NThreads]T
   for i := range channels {
      channels[i].Start(i+1, &channels[(i+1)%NThreads])
   }

   channels[0].put(n)
   fmt.Println(<-res)
}
